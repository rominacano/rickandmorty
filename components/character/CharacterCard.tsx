import { FC, useState } from "react";
import { Button, Card, Col, Grid, Row, Text } from "@nextui-org/react";
import { format } from "date-fns";
import { ICharacter } from "../../interfaces";
import { localFavorite } from "../../utils";

interface Props {
  character: ICharacter;
}

export const CharacterCard: FC<Props> = ({ character }) => {
  const { id, name, species, status, created, image } = character;
  const [isInFavorites, setIsInFavorites] = useState(
    localFavorite.existCharacterInFavorites(character.id)
  );

  const onToggleFavorite = (id: number) => {
    localFavorite.toggleFavorite(id);
    setIsInFavorites(!isInFavorites);
  };

  return (
    <Grid key={id} xs={4}>
      <Card isHoverable css={{ w: "100%", h: "400px" }}>
        <Card.Header css={{ position: "absolute", zIndex: 1, top: 5 }}>
          <Col>
            <Text size={12} weight="bold" transform="uppercase" color="#9E9E9E">
              {species}
            </Text>
            <Text h3 color="black">
              {name}
            </Text>
          </Col>
        </Card.Header>
        <Card.Body css={{ p: 0 }}>
          <Card.Image
            src={image}
            width="100%"
            height="100%"
            objectFit="cover"
            alt={name}
          />
        </Card.Body>
        <Card.Footer
          isBlurred
          css={{
            position: "absolute",
            bgBlur: "#ffffff66",
            borderTop: "$borderWeights$light solid rgba(255, 255, 255, 0.2)",
            bottom: 0,
            zIndex: 1,
          }}
        >
          <Row>
            <Col>
              <Text color="#000" size={14}>
                {`Estado: ${status}`}
              </Text>
              <Text color="#000" size={12}>
                {`Fecha de creaci\u00F3n: ${
                  created && format(new Date(created), "dd/MM/yyyy HH:mm")
                }`}
              </Text>
            </Col>
            <Col>
              <Row justify="flex-end">
                <Button
                  ghost={!isInFavorites}
                  flat
                  auto
                  rounded
                  color="secondary"
                  onClick={() => onToggleFavorite(id)}
                  aria-label="Favorite"
                >
                  <Text
                    css={{ color: "inherit" }}
                    size={12}
                    weight="bold"
                    transform="uppercase"
                  >
                    {isInFavorites ? "En favorito" : "Guardar en favoritos"}
                  </Text>
                </Button>
              </Row>
            </Col>
          </Row>
        </Card.Footer>
      </Card>
    </Grid>
  );
};
