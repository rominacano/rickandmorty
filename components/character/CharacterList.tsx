import { FC } from "react";
import { Grid } from "@nextui-org/react";
import { CharacterCard } from ".";
import { ICharacter } from "../../interfaces";

interface Props {
  characters: ICharacter[];
}

export const CharacterList: FC<Props> = ({ characters }) => {
  return (
    <Grid.Container gap={2} justify="flex-start">
      {characters.map((character) => (
        <CharacterCard key={character.id} character={character} />
      ))}
    </Grid.Container>
  );
};
