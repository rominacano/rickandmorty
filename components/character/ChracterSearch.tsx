import { useState } from "react";
import { useRouter } from "next/router";
import { Collapse, Container, Input, Radio, Spacer } from "@nextui-org/react";

export const ChracterSearch = () => {
  const { push } = useRouter();
  const [searchTerm, setSearchTerm] = useState("");
  const [status, setStatus] = useState("");

  const onSearchTerm = () => {
    if (searchTerm.trim().length === 0) return;
    const query = `name=${searchTerm}&status=${status}`;
    push(`/search/${query}`);
  };

  const onFilter = (value: string) => {
    setStatus(value);
    const query = `name=${searchTerm}&status=${value}`;
    push(`/search/${query}`);
  };

  return (
    <Container css={{ display: "block", justifyContent: "center", mt: 90 }}>
      <Input
        bordered
        placeholder="Buscar personaje"
        width="100%"
        value={searchTerm}
        onChange={(e) => setSearchTerm(e.target.value)}
        onKeyUp={(e) => (e.key === "Enter" ? onSearchTerm() : null)}
      />
      <Spacer x={2} />
      <Collapse bordered title="Estado" subtitle="Filtro por estado">
        <Radio.Group
          defaultValue=""
          css={{ p: 10 }}
          onChange={(value) => onFilter(value)}
        >
          <Radio value="Alive" isSquared>
            Alive
          </Radio>
          <Radio value="Dead" isSquared>
            Dead
          </Radio>
          <Radio value="unknown" isSquared>
            Unknown
          </Radio>
        </Radio.Group>
      </Collapse>
    </Container>
  );
};
