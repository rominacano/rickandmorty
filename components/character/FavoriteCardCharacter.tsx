import { Grid, Card, Col, Row, Text, Button, Tooltip } from "@nextui-org/react";
import { FC } from "react";
import { useCharacter } from "../../hooks";

interface Props {
  characterId: number;
}

export const FavoriteCardCharacter: FC<Props> = ({ characterId }) => {
  const { character } = useCharacter(`character/${characterId}`);

  return (
    <Grid xs={2} key={character.id}>
      <Card isPressable>
        <Card.Body css={{ p: 0 }}>
          <Card.Image
            src={character.image}
            objectFit="cover"
            width="100%"
            height={140}
            alt={character.name}
          />
        </Card.Body>
        <Card.Footer css={{ justifyItems: "flex-start" }}>
          <Row wrap="wrap" justify="space-between" align="center">
            <Text b>{character.name}</Text>
            <Text
              css={{
                color: "$accents7",
                fontWeight: "$semibold",
                fontSize: "$sm",
              }}
            >
              {character.species}
            </Text>
          </Row>
        </Card.Footer>
      </Card>
    </Grid>
  );
};
