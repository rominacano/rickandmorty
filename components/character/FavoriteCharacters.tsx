import { FC } from "react";
import { Grid, Text } from "@nextui-org/react";
import { FavoriteCardCharacter } from "./";

interface Props {
  characters: number[];
}

export const FavoriteCharacters: FC<Props> = ({ characters }) => {
  return (
    <>
      <Text h2 css={{ m: 10 }}>
        Personajes favoritos
      </Text>
      <Grid.Container gap={2} direction="row" justify="flex-start">
        {characters.map((id) => (
          <FavoriteCardCharacter key={id} characterId={id} />
        ))}
      </Grid.Container>
    </>
  );
};
