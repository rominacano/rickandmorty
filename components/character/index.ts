export { CharacterCard } from "./CharacterCard";
export { CharacterList } from "./CharacterList";
export { ChracterSearch } from "./ChracterSearch";
export { FavoriteCardCharacter } from "./FavoriteCardCharacter";
export { FavoriteCharacters } from "./FavoriteCharacters";
