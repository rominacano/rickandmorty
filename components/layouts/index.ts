export { MainLayout } from "./MainLayout";
export { InternalLayout } from "./InternalLayout";
