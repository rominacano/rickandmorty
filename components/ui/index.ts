export { Navbar } from "./Navbar";
export { NoFavorites } from "./NoFavorites";
export { FullScreenLoading } from "./FullScreenLoading";
