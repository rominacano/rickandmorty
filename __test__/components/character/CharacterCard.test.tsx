import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { CharacterCard } from "../../../components/character";
import {
  ICharacter,
  Species,
  Status,
} from "../../../interfaces/character-list";

describe("Character Card", () => {
  const character: ICharacter = {
    id: 218,
    name: "Mechanical Rick",
    status: Status.Alive,
    species: Species.Human,
    image: "https://rickandmortyapi.com/api/character/218",
    created: "2017-12-30T14:33:16.920Z",
  };

  test("renders homepage unchanged", () => {
    const { container } = render(<CharacterCard character={character} />);
    expect(container).toMatchSnapshot();
  });

  test("show name in character card", () => {
    render(<CharacterCard character={character} />);
    expect(screen.getByAltText(character.name)).toBeInTheDocument();
  });
});
