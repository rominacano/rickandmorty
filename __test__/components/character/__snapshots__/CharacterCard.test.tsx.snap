// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Card renders homepage unchanged 1`] = `
<div>
  <div
    class="nextui-c-kRHeuF nextui-c-kRHeuF-ifXZSqS-css nextui-grid-item xs"
  >
    <div
      class="nextui-c-BDLTQ nextui-c-jMTBsW nextui-c-PJLV nextui-c-BDLTQ-jyIjWj-variant-shadow nextui-c-BDLTQ-fmlsco-borderWeight-light nextui-c-BDLTQ-cuwTXc-disableAnimation-false nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-c-BDLTQ-ibAXaqy-css"
      role="section"
      tabindex="-1"
    >
      <div
        class="nextui-c-jxECYO nextui-c-jxECYO-icwRywK-css"
      >
        <div
          class="nextui-c-cMTZRp nextui-c-cMTZRp-ijmTDtb-css"
        >
          <p
            class="nextui-c-PJLV nextui-c-PJLV-hyvuql-weight-bold nextui-c-PJLV-ibGzHov-css"
          >
            Human
          </p>
          <h3
            class="nextui-c-PJLV nextui-c-PJLV-ijtkusk-css"
          >
            Mechanical Rick
          </h3>
        </div>
      </div>
      <div
        class="nextui-c-eFfoBo nextui-c-eFfoBo-ifBQDyP-css"
      >
        <div
          class="nextui-c-iVzbCs nextui-c-iVzbCs-bDGmTT-ready-true nextui-c-iVzbCs-ihinyfY-css nextui-image-container nextui-image--loading"
          data-state="loading"
        >
          <div
            class="nextui-c-FIiRH nextui-c-FIiRH-ibDGmTT-css nextui-image-skeleton"
          />
          <img
            alt="Mechanical Rick"
            class="nextui-c-kbhVdb nextui-c-kbhVdb-ieCSSFU-css nextui-image"
            data-state="loading"
            height="100%"
            src="https://rickandmortyapi.com/api/character/218"
            width="100%"
          />
        </div>
      </div>
      <div
        class="nextui-c-cakKrd nextui-c-cakKrd-gjCMAI-isBlurred-true nextui-c-cakKrd-ijEqIDO-css"
      >
        <div
          class="nextui-c-dfbgCO nextui-c-dfbgCO-fGHEql-fluid-true nextui-c-dfbgCO-idIWWzx-css"
        >
          <div
            class="nextui-c-cMTZRp nextui-c-cMTZRp-ijmTDtb-css"
          >
            <p
              class="nextui-c-PJLV nextui-c-PJLV-iiWWerx-css"
            >
              Estado: Alive
            </p>
            <p
              class="nextui-c-PJLV nextui-c-PJLV-ifpbLBb-css"
            >
              Fecha de creaci√≥n: 30/12/2017 11:33
            </p>
          </div>
          <div
            class="nextui-c-cMTZRp nextui-c-cMTZRp-ijmTDtb-css"
          >
            <div
              class="nextui-c-dfbgCO nextui-c-dfbgCO-fGHEql-fluid-true nextui-c-dfbgCO-iipPaye-css"
            >
              <button
                aria-label="Favorite"
                class="nextui-c-iWjDFM nextui-c-PJLV nextui-c-ldmKBv nextui-c-iWjDFM-iaomLu-bordered-true nextui-c-iWjDFM-wUUIG-color-secondary nextui-c-iWjDFM-gkIlwi-size-md nextui-c-iWjDFM-cwXrJp-borderWeight-normal nextui-c-iWjDFM-cyFBRX-flat-true nextui-c-iWjDFM-exaRme-auto-true nextui-c-iWjDFM-hzYxzX-rounded-true nextui-c-iWjDFM-gglxka-cv nextui-c-iWjDFM-iHRvQZ-cv nextui-c-iWjDFM-gBzeNH-cv nextui-c-iWjDFM-euJLgm-cv nextui-c-PJLV-gulvcB-isFocusVisible-false nextui-c-iWjDFM-iPJLV-css nextui-button nextui-button--ready"
                data-state="ready"
                type="button"
              >
                <span
                  class="nextui-button-text"
                >
                  <p
                    class="nextui-c-PJLV nextui-c-PJLV-hyvuql-weight-bold nextui-c-PJLV-ikBSpPJ-css"
                  >
                    Guardar en favoritos
                  </p>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
